### PR TITLE
Issue #2870 Honour KUBECONFIG env if set

### LINKED ISSUE #2870: Minishift needs to honour KUBECONFIG
ATM `minishift start` does not honour the _KUBECONFIG_ environment variable. For example,

```
$ export KUBECONFIG=/Users/john/tmp/minishift-kube-config
$ minsihift start
```

The certificate data as well as context for system:admin gets still written into _~/.kube/config_. This prevents me from better controlling my various OpenShift configurations.

### PR DESCRIPTION
**Pull request by praveenkumar**
Fixes #2870

cc @hferentschik 

Steps to test the pull request

```
$ export KUBECONFIG=/tmp/foo/config
$ minishift start
$ cat /tmp/foo/config
$ oc whoami <= on the same terminal where you have exported the KUBECONFIG file.
```

### CODE REVIEW

**Comment by praveenkumar** [comment]
@LalatenduMohanty this PR is only for honor the KUBECONFIG env nothing more. what you suggested is something part of #2871. No 'MINISHIFT_KUBECONFIG` and `KUBECONFIG` 2 different thing and all the folks who works around kubernetes known to `KUBECONFIG`.

**Comment by LalatenduMohanty** [comment]
> No `MINISHIFT_KUBECONFIG` and `KUBECONFIG` 2 different thing and all the folks who works around kubernetes known to `KUBECONFIG`.

You are right about `KUBECONFIG`. I missed it https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#set-the-kubeconfig-environment-variable

File: pkg/minishift/oc/oc.go [hunk 1]
```diff
@@ -226,3 +225,21 @@ func mergeKubeConfigs(configs []*clientcmdapi.Config) *clientcmdapi.Config {
 	}
 	return mergedConfig
 }
+
+func getKubeConfigGlobalPath() (string, error) {
+	kubeConfigGlobalPath := os.Getenv("KUBECONFIG")
+	if kubeConfigGlobalPath != "" {
+		kubeConfigGlobalPath = strings.FieldsFunc(kubeConfigGlobalPath, splitKubeConfig)[0]
```
Comment (hferentschik) [inline pkg/minishift/oc/oc.go]: So the idea is to take the first config in the path, right? Might be worth to document, probably in a godoc:
"getKubeConfigGlobalPath returns the path to the first entry in KUBECONFIG or if not set ~/home/.kube/config" - something along these lines. 

This will help other people to understand the code. I don't think it is so well known that KUBECONFIG can have multiple path elements.

File: pkg/minishift/oc/oc.go [hunk 2]
```diff
@@ -226,3 +225,21 @@ func mergeKubeConfigs(configs []*clientcmdapi.Config) *clientcmdapi.Config {
 	}
 	return mergedConfig
 }
+
+func getKubeConfigGlobalPath() (string, error) {
```
Comment (hferentschik) [inline pkg/minishift/oc/oc.go]: Nice to extract this into a function.

File: pkg/minishift/oc/oc.go [hunk 3]
```diff
@@ -94,16 +94,15 @@ func (oc *OcRunner) AddSudoerRoleForUser(user string) error {
 func (oc *OcRunner) AddSystemAdminEntryToKubeConfig(ocPath string) error {
 	var existingKubeConfig, newKubeConfig *clientcmdapi.Config
 
-	user, err := user.Current()
+	minishiftKubeConfigPath := oc.KubeConfigPath
+	kubeConfigGlobalPath, err := getKubeConfigGlobalPath()
```
Comment (LalatenduMohanty) [inline pkg/minishift/oc/oc.go]: globalKubeConfigPath sounds better as a variable name here.

Comment (LalatenduMohanty) [reply pkg/minishift/oc/oc.go]: Same with function name too i.e. `getGlobalKubeConfigPath`

File: pkg/minishift/oc/oc.go (hunk 1 above)
Comment (praveenkumar) [reply pkg/minishift/oc/oc.go]: > So the idea is to take the first config in the path, right? Might be worth to document, probably in a godoc

@hferentschik Sure that something we can add.

> I don't think it is so well known that KUBECONFIG can have multiple path elements.

https://kubernetes.io/docs/tasks/access-application-cluster/configure-access-multiple-clusters/#set-the-kubeconfig-environment-variable

File: pkg/minishift/oc/oc.go (hunk 3 above)
Comment (praveenkumar) [reply pkg/minishift/oc/oc.go]: @LalatenduMohanty it was there before so I didn't change it but can be updated.

**Comment by LalatenduMohanty** [comment]
So minikube has a function for handling KUBECONFIG but we can not use it in its current form https://github.com/minishift/minishift/blob/master/pkg/minikube/kubeconfig/config.go#L54-L6

File: pkg/minishift/oc/oc_test.go [hunk 4]
```diff
@@ -144,6 +146,24 @@ func TestFlagExist(t *testing.T) {
 	assert.False(t, flagExist(expectedOptions, "proxy"))
 }
 
+func Test_get_kubeconfig_global_path(t *testing.T) {
+	testdata := []struct {
+		KubeConfig string
+		Expected   string
+	}{
+		{KubeConfig: "/tmp/john/config", Expected: "/tmp/john/config"},
```
Comment (hferentschik) [inline pkg/minishift/oc/oc_test.go]: what's about adding some more boundary conditions? What does things like ":/home/john" do or  ";;", etc. The tests are only covering the happy paths.

On another note, maybe the split character should be determined at runtime based on the OS, aka ":" on darwin/linux and ';' on windows. That would be more accurate imo.

**Comment by LalatenduMohanty** [comment]
retest this please

File: pkg/minishift/oc/oc.go [hunk 5]
```diff
@@ -226,3 +225,23 @@ func mergeKubeConfigs(configs []*clientcmdapi.Config) *clientcmdapi.Config {
 	}
 	return mergedConfig
 }
+
+// getGlobalKubeConfigPath returns the path to the first entry in KUBECONFIG environment variable
+// or if KUBECONFIG not set then ~/home/.kube/config
+func getGlobalKubeConfigPath() (string, error) {
```
Comment (hferentschik) [inline pkg/minishift/oc/oc.go]: For what it's worth, why getGlobalKubeConfigPath? What's global about this? Is not a simple (get)KubeConfigPath more fitting?

**Comment by praveenkumar** [comment]
retest this please.

File: pkg/minishift/oc/oc.go [hunk 6]
```diff
@@ -216,3 +216,27 @@ func mergeKubeConfigs(configs []*clientcmdapi.Config) *clientcmdapi.Config {
...
+		usr, err := user.Current()
+		if err != nil {
+			return "", fmt.Errorf("unable to find current user: %v", err)
+		}
+		globalKubeConfigPath = filepath.Join(usr.HomeDir, ".kube", "config")
+	}
+	return globalKubeConfigPath, nil
+}
+
+func splitKubeConfig(r rune) bool {
```
Comment (hferentschik) [inline pkg/minishift/oc/oc.go]: +1, I think this is more accurate

File: pkg/minishift/oc/oc_test.go [hunk 7]
```diff
@@ -128,6 +131,42 @@ func TestFlagExist(t *testing.T) {
...
 
+func Test_get_kubeconfig_global_path(t *testing.T) {
+	type testdata struct {
+		KubeConfig string
+		Expected   string
+	}
+	var tests []testdata
+
+	if runtime.GOOS == "windows" {
+		tests = []testdata{{KubeConfig: "/tmp/john/config", Expected: "/tmp/john/config"},
```
Comment (hferentschik) [inline pkg/minishift/oc/oc_test.go]: +1, better coverage

File: pkg/minishift/oc/oc.go (hunk 5 above)
Comment (hferentschik) [reply pkg/minishift/oc/oc.go]: The godocs are still wrong. Either you use $HOME/.kube/config or ~/.kube/config. '~/home' is wrong.

**Comment by praveenkumar** [comment]
> What's about any update to the user docs? Would it not make sense to add something there as well. It might be interesting for the user to which file the minishift config gets written.

@hferentschik Users who actually going to use this, already well aware of how the `KUBECONFIG` variable is used and configs are written.

> Maybe we should also log it somewhere?

log it you mean print it on console that `config written to <kube_config> path` ?

**Comment by hferentschik** [comment]
> Users who actually going to use this, already well aware of how the KUBECONFIG variable is used and configs are written.

So you are saying that it is common knowledge that new cluster configuration is added to the first config found in the path? To me that was new. Personally, I think it is better to err on being a bit more verbose. Anyways,up to you.

> log it you mean print it on console that config written to <kube_config> path 

Maybe. I guess one would have to see how it looks like. It might be quite useful actually. What I initially had in mind is that at least we write some debug trace in an appropriate log level.

**Comment by praveenkumar** [comment]
> Maybe. I guess one would have to see how it looks like. It might be quite useful actually. What I initially had in mind is that at least we write some debug trace in an appropriate log level

@hferentschik That I added as glog level 2 for debug purpose, so we are good now.

File: pkg/minishift/oc/oc.go [hunk 8]
```diff
@@ -92,24 +94,25 @@ func (oc *OcRunner) AddSudoerRoleForUser(user string) error {
...
-		return fmt.Errorf("unable to find current user: %v", err)
+		return err
 	}
-
-	minishiftKubeConfigPath := oc.KubeConfigPath
-	kubeConfigGlobalPath := filepath.Join(user.HomeDir, ".kube", "config")
-
-	if err := os.MkdirAll(filepath.Join(user.HomeDir, ".kube"), 0755); err != nil {
-		return fmt.Errorf("unable to create kubeconfig dir ($HOME/.kube): %v", err)
+	if glog.V(2) {
```
Comment (hferentschik) [inline pkg/minishift/oc/oc.go]: :-)

File: pkg/minishift/oc/oc.go [hunk 9]
```diff
@@ -216,3 +219,29 @@ func mergeKubeConfigs(configs []*clientcmdapi.Config) *clientcmdapi.Config {
 	}
 	return mergedConfig
 }
+
+// getGlobalKubeConfigPath returns the path to the first entry in KUBECONFIG environment variable
+// or if KUBECONFIG not set then $HOME/.kube/config
```
Comment (hferentschik) [inline pkg/minishift/oc/oc.go]: :-)
